Decompress gziped body before logging

belimawr opened the pull request:
The request body was logged directly, as raw data, this is problematic if the body is compressed or is not just ASCII data.

This commit fixes it by reading the `Content-Encoding` header and decompressing if it is `gzip`.

Request logging is disabled by default and a flag (`logrequests`) to enable it is added.

leehinman commented:
Fine with me, but lets wait for @mauri870 's opinion.  He added the feature.

mauri870 commented:
I tend to use this feature quite often to debug exporter/receivers and the beats to some extent. I simply set `compression: none` or `compression_level: 0` and then it prints the body properly. I'm fine with just printing a placeholder or warning if the request is compressed, but I would rather keep printing the uncompressed body.

mauri870 commented on cmd/mock-es/main.go:
```diff
@@ -127,7 +127,7 @@ func loggingMiddleware(next http.Handler) http.Handler {
 		}
 
-		log.Printf("%s %s\n%s", r.Method, r.URL.RequestURI(), body)
+		log.Printf("%s %s\n", r.Method, r.URL.RequestURI())
```
I think we can use `http.DetectContentType(body)` here and if the body is text/plain or JSON we log it, otherwise we can print `<compressed>` or something similar.

belimawr commented:
> I tend to use this feature quite often to debug exporter/receivers and the beats to some extent. I simply set `compression: none` or `compression_level: 0` and then it prints the body properly. I'm fine with just printing a placeholder or warning if the request is compressed, but I would rather keep printing the uncompressed body.

I see it now. I was really wondering what was the purpose of logging the whole body, specially because I was just running Filebeat and it was breaking my terminal.

I like the idea of keeping it and printing a placehoder when the body is compressed.

belimawr commented:
Folks, I went a bit further and decompressed the body before logging. I also added  a flag to enable the request logging, so if someone (like me) is interested only in the metrics, the logging ca be disabled.

I also added a couple of log entries to indicate the server is about to start.

leehinman commented on cmd/mock-es/main.go:
```diff
@@ -47,6 +50,7 @@ func init() {
 	flag.StringVar(&certFile, "certfile", "", "path to PEM certificate file, empty sting is no TLS")
 	flag.StringVar(&keyFile, "keyfile", "", "path to PEM private key file, empty sting is no TLS")
 	flag.DurationVar(&delay, "delay", 0, "Go 'time.Duration' to wait before processing API request, 0 is no delay")
+	flag.BoolVar(&logRequests, "logrequests", false, "Enable request logging")
```
Proposal:  How about changing this to "verbose", and then both the "starting HTTP server" and the body logging could be enabled or disabled with that flag?

I like being able to keep the server silent, or silent with metrics I can pipe to jq.

belimawr replied:
That makes more sense, I'll update the PR

leehinman reviewed (approved):
LGTM
